cmd/dockerd/hack/malformed_host_override_test.go

```diff
@@ -33,27 +41,32 @@ func TestHeaderOverrideHack(t *testing.T) {
 			[]byte("GET /foo\nFoo: Bar\nHost: /var/run/docker.sock\nUser-Agent: Docker\r\n\r\n"),
 			[]byte("GET /foo\nFoo: Bar\nHost: /var/run/docker.sock\nUser-Agent: Docker\r\n\r\n"),
 		},
+		{ // Test for https://github.com/docker/docker/issues/23045
+			[]byte("GET /foo\nUser-Agent: Docker\r\n\r\n" + strings.Repeat("test", 15000)),
+			[]byte("GET /foo\nUser-Agent: Docker\r\n\r\n" + strings.Repeat("test", 15000)),
+		},
 	}
-	l := MalformedHostHeaderOverrideConn{client, true}
-	read := make([]byte, 4096)
+
+	h0 := "GET /foo\nUser-Agent: Docker\r\n\r\n"
+	h0 = h0 + strings.Repeat("a", 4096-len(h0)-1) + "\n"
+	tests = append(tests, [2][]byte{[]byte(h0), []byte(h0)})
```

Comment:
More precisely, if '\n' appears in the last 7 bytes of the first 4096 bytes, the `index out of range` error will be triggered because `i+n` will be greater than 4095.